src/SignalR/common/Shared/AsyncEnumerableAdapters.cs

```diff
@@ -13,9 +13,24 @@ namespace Microsoft.AspNetCore.SignalR.Internal
     // True-internal because this is a weird and tricky class to use :)
     internal static class AsyncEnumerableAdapters
     {
-        public static IAsyncEnumerable<object?> MakeCancelableAsyncEnumerable<T>(IAsyncEnumerable<T> asyncEnumerable, CancellationToken cancellationToken = default)
+        public static async IAsyncEnumerable<object?> MakeCancelableAsyncEnumerable<T>(IAsyncEnumerable<T> asyncEnumerable, [EnumeratorCancellation] CancellationToken cancellationToken = default)
```

Comment:
If I recall this issue correctly, the idea was that we could avoid the allocation of a new `IAsyncEnumerable<object?>` altogether. This just changes it from a CancelableAsyncEnumerable allocation to an allocation of the [compiler-generated iterator](https://docs.microsoft.com/en-us/archive/msdn-magazine/2019/november/csharp-iterating-with-async-enumerables-in-csharp-8#under-the-hood-of-async-iterators) backing this async iterator method.

The idea behind #7960 is that we can just cast to `(IAsyncEnumerator<object?>)` without allocating at all if `!typeof(T).IsValueType` like I did [here in the original PR](https://github.com/dotnet/aspnetcore/pull/6791/commits/4de77e9c2c5f4cc9a18b1efcf4bf7906973861b1#diff-feae18e7b3265bffa80aab26f8dc8f8111684fda1e1b2781e775d5f69dc11851) before @davidfowl asked me to switch to a foreach and then switch it back after merging 😜.

Reply:
Got it, I will update the PR accordingly.

Reply:
So let me reiterate my current understanding. We would need to change `IAsyncEnumerable<object?> 
 MakeCancelableAsyncEnumerable` to `IAsyncEnumerator<object?> 
 MakeCancelableAsyncEnumerator` so by not having the enumerable but the enumerator, we avoid the allocation of the enumerable object for reference types.
For value types, we would still need to keep a `BoxedAsyncEnumerator`.

Because how `HubMethodDescriptor` uses these adapter methods, will need to update the `Channel<T>` signature and [implementation](https://github.com/dotnet/aspnetcore/blob/main/src/SignalR/common/Shared/AsyncEnumerableAdapters.cs#L26) too. If I see this correctly, for `Channel<T>` we will need to re-introduce a `ChannelAsyncEnumerator`  (like the one you had) probably for both netcore and netstandard implementation.

Then we would need to refactor [HubMethodDescriptor](https://github.com/dotnet/aspnetcore/blob/main/src/SignalR/server/Core/src/Internal/HubMethodDescriptor.cs#L121) to generate proper `Func<>` and then iterate using these iterators in [DefaultHubDispatcher](https://github.com/dotnet/aspnetcore/blob/main/src/SignalR/server/Core/src/Internal/DefaultHubDispatcher.cs#L478)

Is my understanding correct?
Is this the intended outcome the issue?

Reply:
@halter73 @BrennanConroy or @davidfowl could you advise on the above questions?

Reply:
Your understanding sounds right to me! We'll need to keep a `BoxedAsyncEnumerator`. I think that a lot of the changes in https://github.com/dotnet/aspnetcore/pull/6791/commits/41544a289a99c9950b260beee329600b61dd8441 will need to be undone. Does that answer your questions?

Reply:
Yes, thank you.

Reply:
I have updated the PR as discussed.